tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,253 @@
+	for {
+		m := tb.messageReceiver.Receive()
+
+		if err := tb.handleMessage(m); err != nil {
+			return err
+		}
+	}
+}
+
+// listens to new layers.
+func (tb *TortoiseBeacon) tickLoop() {
+	for {
+		select {
+		case layer := <-tb.layerTicker:
+			go tb.onTick(layer)
+		case <-tb.CloseChannel():
+			return
+		}
+	}
+}
+
+// the logic that happens when a new layer arrives.
+// this function triggers the start of new CPs.
+func (tb *TortoiseBeacon) onTick(layer types.LayerID) {
+	tb.layerMu.Lock()
+	if layer > tb.lastLayer {
+		tb.lastLayer = layer
+	}
+
+	tb.layerMu.Unlock()
+
+	epoch := layer.GetEpoch()
+	tb.Debug("tortoise beacon got tick at layer %v epoch %v, sleeping for %v", layer, epoch, tb.networkDelta)
+
+	if layer.GetEpoch().IsGenesis() {
+		tb.With().Info("not starting tortoise beacon since we are in genesis epoch", layer)
+		return
+	}
+
+	tb.beaconsMu.Lock()
+
+	if _, ok := tb.beacons[epoch]; ok {
+		tb.beaconsMu.Unlock()
+
+		// Already handling this epoch
+		return
+	}
+
+	tb.beacons[epoch] = make(chan types.Hash32, 1)
+	tb.beaconsMu.Unlock()
+
+	go func() {
+		tb.roundTicker(epoch)
+
+		tb.timelyProposalsMu.Lock()
+		beacon := tb.beaconCalculator.CalculateBeacon(tb.timelyProposals)
+		tb.timelyProposalsMu.Unlock()
+
+		tb.beaconsMu.Lock()
+		tb.beacons[epoch] <- beacon
+		close(tb.beacons[epoch])
+		tb.beaconsMu.Unlock()
+	}()
+
+	atxList := tb.atxDB.GetEpochAtxs(epoch - 1)
+	m := NewMessage(epoch, 0, atxList)
+	tb.messageSender.Send(m)
+}
+
+func (tb *TortoiseBeacon) handleMessage(m Message) error {
+	epoch := m.Epoch()
+	tb.classifyMessage(m, epoch)
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) classifyMessage(m Message, epoch types.EpochID) {
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	pair := EpochRoundPair{
+		EpochID: epoch,
+		Round:   currentRound,
+	}
+
+	if m.Round() <= currentRound-1 {
+		tb.timelyProposalsMu.Lock()
+		tb.timelyProposals[pair] = m.Payload()
+		tb.timelyProposalsMu.Unlock()
+	} else if m.Round() == currentRound-2 {
+		tb.delayedProposalsMu.Lock()
+		tb.delayedProposals[pair] = m.Payload()
+		tb.delayedProposalsMu.Unlock()
+	} else {
+		tb.lateProposalsMu.Lock()
+		tb.lateProposals[pair] = m.Payload()
+		tb.lateProposalsMu.Unlock()
+	}
+}
+
+func (tb *TortoiseBeacon) roundTicker(epoch types.EpochID) {
+	ticker := time.NewTicker(tb.networkDelta)
+	defer ticker.Stop()
+
+	tb.currentRoundsMu.Lock()
+	tb.currentRounds[epoch] = 0
+	tb.currentRoundsMu.Unlock()
+
+	for i := 1; i < tb.lastPossibleRound(); i++ {
+		select {
+		case <-ticker.C:
+			tb.currentRoundsMu.Lock()
+			tb.currentRounds[epoch]++
+			round := tb.currentRounds[epoch]
+			tb.currentRoundsMu.Unlock()
+
+			pair := EpochRoundPair{
+				EpochID: epoch,
+				Round:   round - 1, // proposals from the previous round are needed
+			}
+
+			tb.timelyProposalsMu.Lock()
+			proposals := tb.timelyProposals[pair]
+			tb.timelyProposalsMu.Unlock()
+
+			m := NewMessage(epoch, round, proposals)
```

Comment:
in this implementation, on each round, a miner will send all the ATX she thinks should be in this epochs beacon
this is very network intensive, that is why the protocol mentions 2 kinds of messages, since the initial proposed set is sent, all later messages vote for a set by voting on a message